[FLINK-7638] [flip6] Port CurrentJobsOverviewHandler to new REST endpoint

tillrohrmann opened the pull request:
## What is the purpose of the change

Ports the CurrentJobsOverviewHandler to the new REST endpoint by letting it implement
the LegacyRestHandler interface. This commit changes the JobDetails JSON such that it
now contains the number of tasks for each ExecutionState, including SCHEDULED,
DEPLOYING, CREATED and RECONCILING. These state will now also be displayed in the
web frontend.

## Verifying this change

- Have tested manually that the old web frontend is still working
- Added `JobDetailsTest` and `MultipleJobsDetailsTest` which test the un/marshalling

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

tzulitai commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:
```diff
@@ -258,6 +260,23 @@ public void start() throws Exception {
+
+		return combinedJobDetails.thenApply(
+			(Collection<JobDetails> jobDetails) ->
+				new MultipleJobsDetails(jobDetails.toArray(new JobDetails[numberJobsRunning]), null));
```
Can we somehow lessen the array <-> collection conversions here?
For example, I think in `FutureUtils.combineAll` the completed results were actually originally an array before they were returned as a collection, and then here we're converting it back to an array again.

tillrohrmann replied:
`FutureUtils.combineAll` returns a `Arrays.ArrayList`. This type is directly backed by the provided array. When calling `Arrays.ArrayList#toArray`, this array will be copied. Thus, I can change `MultipleJobsDetails` to store the individual `JobDetails` as a list. Then there should be no copy involved.

tzulitai commented on flink-runtime/src/main/java/org/apache/flink/runtime/messages/webmonitor/JobDetails.java:
```diff
@@ -145,8 +180,89 @@ public String toString() {
+
+			for (ExecutionState executionState : ExecutionState.values()) {
+				jsonGenerator.writeNumberField(executionState.name().toLowerCase(), perState[executionState.ordinal()]);
+			}
```
I guess this is the bit which requires this nasty custom serialization / deserialization.

I wonder if it is possible to have a proper POJO w/ Jackson annotation for this. For the `JobDetails` class we would only need an additional constructor (without `@JsonCreator` annotation) that accepts the "awkward" `int[] perState` and converts it to a `List<PerExecutionState>`, `PerExecutionState` being a newly introduced POJO to hold per-state.

If that's possible things would be much cleaner here.

tillrohrmann replied:
I think it's actually alright to have a custom serialization/deserialization logic defined if needed. I think this should be more efficient than the generic serialization because you don't have to use reflection magic. Moreover, it's encapsulated in the `JobDetails class. Thus, if need is there, then this could also be changed as you've proposed.

tzulitai commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/CurrentJobsOverviewHandler.java:
```diff
@@ -69,6 +73,13 @@ public CurrentJobsOverviewHandler(
+	public CompletableFuture<MultipleJobsDetails> handleRequest(HandlerRequest<EmptyRequestBody, EmptyMessageParameters> request, DispatcherGateway gateway) {
+		CompletableFuture<MultipleJobsDetails> jobDetailsFuture = gateway.requestJobDetails(timeout);
+
+		return jobDetailsFuture;
```
nit: might as well just make this a one-liner:
`return gateway.requestJobDetails(timeout);`

tillrohrmann replied:
Will change it.

tillrohrmann commented:
Thanks for the review @tzulitai. I've addressed your comments. Concerning the custom serialization logic, I think it does not hurt to have it, because it should be more efficient than the generic serialization logic. Moreover it's an implementation detail which could be changed later on as well. Given that, I would like to keep it at the moment like this.

tzulitai commented:
@tillrohrmann Yes that makes sense, lets do that separately in the future then.

+1.

tzulitai commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/CurrentJobsOverviewHandler.java:
```diff
@@ -92,24 +101,26 @@ public CurrentJobsOverviewHandler(
+								jobDetailsSerializer.serialize(detail, gen, null);
 							}
 							gen.writeEndArray();
 						} else {
```
On a second look, I think this response variant is not covered by the ported handler.

tillrohrmann replied:
Yes, because I think it's not so nice to return different answer formats depending on how the handler was instantiated. The respective information can be easily extracted from the complete overview. Moreover, I think we never used these endpoints in the first place.

tzulitai commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/CurrentJobsOverviewHandler.java:
```diff
@@ -92,24 +101,26 @@ public CurrentJobsOverviewHandler(
 						gen.writeStartObject();
 
 						if (includeRunningJobs && includeFinishedJobs) {
 							gen.writeArrayFieldStart("running");
```
"running" --> can use `MultipleJobsDetails.FIELD_NAME_JOBS_RUNNING`

tillrohrmann replied:
True, will change it.

tzulitai commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/CurrentJobsOverviewHandler.java:
```diff
@@ -92,24 +101,26 @@ public CurrentJobsOverviewHandler(
 							}
 							gen.writeEndArray();
 
 							gen.writeArrayFieldStart("finished");
```
"finished" --> can use `MultipleJobsDetails.FIELD_NAME_JOBS_FINISHED`

tillrohrmann replied:
Will change it.

tillrohrmann commented:
Thanks for your review @tzulitai. Travis passed. Will merge the PR.
